### PR TITLE
Upcoming [DI-19807] : Added CreateAlertDefinition workflow

### PR DESCRIPTION
**Pull request by santoshp210-akamai**
## Description 📝
Highlight the Pull Request's context and intentions.

## Changes  🔄
List any change relevant to the reviewer.
- ...
- ...

## Target release date 🗓️
Please specify a release date to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| 📷 | 📷 |

### Prerequisites
(How to setup test environment)
- ...
- ...

### Reproduction steps
(How to reproduce the issue, if applicable)
- ...
- ...

### Verification steps
(How to verify changes)
- ...
- ...

## As an Author I have considered 🤔

*Check all that apply*

- [ ] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [ ] ➕ Adding a changeset
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

---
## Commit message and pull request title format standards

> **Note**: Remove this section before opening the pull request
**Make sure your PR title and commit message on squash and merge are as shown below**

`<commit type>: [JIRA-ticket-number] - <description>`

**Commit Types:**
- `feat`: New feature for the user (not a part of the code, or ci, ...).
- `fix`: Bugfix for the user (not a fix to build something, ...).
- `change`: Modifying an existing visual UI instance. Such as a component or a feature.
- `refactor`: Restructuring existing code without changing its external behavior or visual UI. Typically to improve readability, maintainability, and performance.
- `test`: New tests or changes to existing tests. Does not change the production code.
- `upcoming`: A new feature that is in progress, not visible to users yet, and usually behind a feature flag.

**Example:** `feat: [M3-1234] - Allow user to view their login history`

---

### CODE REVIEW

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx
Comment (satkumar-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx:1]: Alerts Landing page should not be available while creating Alert definition.

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertsLanding.tsx
Comment (satkumar-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertsLanding.tsx:1]: move the Notify component to a separate component

File: packages/manager/src/features/CloudPulse/Alerts/CreateAlert/CreateAlertDefinition.tsx
Comment (satkumar-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/CreateAlert/CreateAlertDefinition.tsx:1]: fix the submit and cancel button placement. Placement should be from the right side of the screen and should avoid adding un-necessarily extra space.

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx
Comment (santoshp210-akamai) [reply packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx:1]: Yeah Jay, I've fixed it with a different component for AlertDefinition

File: packages/api-v4/src/cloudpulse/alerts.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+export const createAlertDefinition = (data: CreateAlertDefinitionPayload) =>
+  Request<Alert>(
+    setURL(
+      // `http://blr-lhvlls.bangalore.corp.akamai.com:9000/v4/simple/monitor/alerts`
+      `${API_ROOT}/monitor/alerts`
+    ),
+    setMethod('POST'),
+    setHeaders({
+      Authorization: 'Bearer vagrant',
```
Comment (kdudeja-akamai) [inline packages/api-v4/src/cloudpulse/alerts.ts]: this authorisation header is hardcoded...please fix it

File: packages/api-v4/src/cloudpulse/alerts.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// import { API_ROOT } from 'src/constants';
+import { createAlertDefinitionSchema } from '@linode/validation';
+import Request, { setURL, setMethod, setData, setHeaders } from '../request';
+import { Alert, CreateAlertDefinitionPayload } from './types';
+import { API_ROOT } from 'src/constants';
+// import { createAlertDefinitionSchema } from '@linode/validation';
```
Comment (kdudeja-akamai) [inline packages/api-v4/src/cloudpulse/alerts.ts]: remove commented code from every file

File: packages/api-v4/src/cloudpulse/services.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    setMethod('GET'),
+    setHeaders({
+      Authorization: 'Bearer vagrant',
+    })
+  );
+
+export const getMetricDefinitionsByServiceType = (serviceType: string) => {
+  return Request<Page<MetricDefinitions>>(
+    setURL(
+      // `https://blr-lhv95n.bangalore.corp.akamai.com:9000/v4/monitor/services/${serviceType}/metric-definitions`
```
Comment (kdudeja-akamai) [inline packages/api-v4/src/cloudpulse/services.ts]: remove commented code

File: packages/api-v4/src/cloudpulse/services.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      `${API_ROOT}/monitor/services/${serviceType}/metric-definitions`
+    ),
+    setMethod('GET'),
+    setHeaders({
+      Authorization: 'Bearer vagrant',
```
Comment (kdudeja-akamai) [inline packages/api-v4/src/cloudpulse/services.ts]: remove hardcoding

File: packages/api-v4/src/cloudpulse/services.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import { ResourcePage as Page } from 'src/types';
+
+export const getCloudViewServiceTypes = () =>
+  Request<ServiceTypes>(
+    setURL(
+      // `https://blr-lhv95n.bangalore.corp.akamai.com:9000/v4/monitor/services`
+      `${API_ROOT}/monitor/services`
+    ),
+    setMethod('GET'),
+    setHeaders({
```
Comment (kdudeja-akamai) [inline packages/api-v4/src/cloudpulse/services.ts]: duplicate lines...you can set headers at once in a function

File: packages/api-v4/src/cloudpulse/types.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+export interface CreateAlertDefinitionPayload {
+  alertName: string | null;
+  region: string | null;
+  description?: string | null;
+  serviceType: string | null;
```
Comment (kdudeja-akamai) [inline packages/api-v4/src/cloudpulse/types.ts]: make sure...fields are aligned with this doc https://docs.google.com/document/d/1csUE83r_y1fnWOflAE9wMXLWuWWJIgWzdFZns3n3Bbk/edit#heading=h.ank5375psbvy...i can see difference in some fields like servicetype

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        path="/monitor/cloudpulse/alerts/definitions/create"
+      />
+    </Switch>
+  );
+};
+
+const AlertDefinitionDetail = () => {
+  return (
+    <Paper>
+      <Typography variant="body1">Alert Definition</Typography>
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx]: as i see, its just a placeholder for now..I would suggest moving it to separate file

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertsLanding.tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+import { Tabs } from 'src/components/Tabs/Tabs';
+
+import { AlertDefinition } from './AlertDefinitions';
+import { Notify } from './Notification';
+import { RecentActivity } from './RecentActivity';
+
+const AlertsLanding = React.memo(() => {
+  const { path } = useRouteMatch();
+  const history = useHistory();
+  const tabs = [
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertsLanding.tsx:25]: we can move tabs into separate function to maintain readability...ideally, try to have small function as far as possible

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/RecentActivity.tsx [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import * as React from 'react';
+
+import { Paper } from 'src/components/Paper';
+import { Typography } from 'src/components/Typography';
+
+export const RecentActivity = () => {
+  return (
+    <Paper>
+      <Typography variant="body1">Recent Activity</Typography>
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/AlertLanding/RecentActivity.tsx:9]: for the pages that are yet to be developed, can we add something like Coming soon later...some understanding for others ..we can discuss

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx:1]: NO unit tests for all 4 files under alertlanding ...are others following the same or we yet to add?

File: packages/manager/src/features/CloudPulse/Alerts/CreateAlert/AddChannelListing.tsx [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import { DeleteOutlineOutlined } from "@mui/icons-material";
+
+interface ChannelListProps {
+  notifications: any[];
+  onChangleNotifications: (notifications: any[]) => void;
+  onClickAddNotification: () => void;
+};
+
+export const AddChannelListing = (props: ChannelListProps) => {
+  const {notifications, onChangleNotifications, onClickAddNotification} = props;
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/CreateAlert/AddChannelListing.tsx]: onChangleNotifications spelling

File: packages/manager/src/features/CloudPulse/Alerts/CreateAlert/AddChannelListing.test.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      type: "Email",
+      values: {
+        to: ["test@example.com"],
+      }
+    },
+];
+  const mockOnChangeNotifications =  vi.fn();
+  const mockOnAddNotification = vi.fn();
+
+  it('should render channels with the notification list', () => {
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/CreateAlert/AddChannelListing.test.tsx:21]: can add some test cases like if notifications are empty and to cover other edge cases.

File: packages/manager/src/features/CloudPulse/Alerts/CreateAlert/CreateAlertDefinition.tsx [hunk 13]
```diff
@@ -0,0 +1,294 @@
+import { APIError } from '@linode/api-v4';
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/CreateAlert/CreateAlertDefinition.tsx:1]: unit test pending...i guess Santosh will push when he's back

File: packages/manager/src/features/CloudPulse/Alerts/CreateAlert/CreateAlertDefinition.tsx [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+};
+
+export const CreateAlertDefinition = React.memo(() => {
+  const [openAddNotification, setOpenAddNotification] = React.useState(false);
+  const [notifications, setNotifications] = React.useState<any>([]);
+  const { mutateAsync } = useCreateAlertDefinition();
+  const { enqueueSnackbar } = useSnackbar();
+  const { data: engineOptions } = useDatabaseEnginesQuery(true); 
+
+  const onSubmitAddNotification = (notification: any) => {
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/CreateAlert/CreateAlertDefinition.tsx:68]: 'any' type is used in many places...i guess it makes code loosely typed..can we make it more specific..let me know if I am wrong here

File: packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Custom/Metrics/DimensionFilterField.tsx [hunk 15]
```diff
@@ -0,0 +1,161 @@
+import { Dimension } from '@linode/api-v4';
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Custom/Metrics/DimensionFilterField.tsx:1]: unit test case pending it seems..please check

File: packages/manager/src/mocks/serverHandlers.ts [hunk 16]
```diff
@@ -548,6 +548,122 @@ const parentAccountNonAdminUser = accountUserFactory.build({
   username: 'NonAdminUser',
 });
 
+const cloudpulse = [
+  http.get('*/monitor/services', () => {
```
Comment (kdudeja-akamai) [inline packages/manager/src/mocks/serverHandlers.ts:552]: any mock response for error?

File: packages/manager/src/features/CloudPulse/CloudPulseTabs.tsx [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+      <TabLinkList tabs={tabs} />
+
+      <React.Suspense fallback={<SuspenseLoader />}>
+        {/* <TabPanels>
+          <SafeTabPanel index={0}>
+            <Switch>
+              <Route
+                component={dashboard}
+                path={`${props.match.url}/dashboards`}
+              />
```
Comment (kdudeja-akamai) [inline packages/manager/src/features/CloudPulse/CloudPulseTabs.tsx]: commented code...please remove if not required

File: packages/validation/src/cloudpulse.schema.ts [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+const content = object({
+  email: string(),
+});
+
+const notifications = object({
+  notification_type: string(),
+  content,
+});
+
+const engionOptionValidation = string().when('serviceType', {
```
Comment (kdudeja-akamai) [inline packages/validation/src/cloudpulse.schema.ts]: typo

**Review by kdudeja-akamai — CHANGES_REQUESTED** [review]

File: packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx (hunk 10 above)
Comment (santoshp210-akamai) [reply packages/manager/src/features/CloudPulse/Alerts/AlertLanding/AlertDefinitions.tsx:1]: Hi Kunal, Under alert landing only the AlertsLanding component is functional . So the test file will be written for only that component.

File: packages/api-v4/src/cloudpulse/alerts.ts (hunk 1 above)
Comment (santoshp210-akamai) [reply packages/api-v4/src/cloudpulse/alerts.ts]: Yeah sure Kunal

File: packages/manager/src/mocks/serverHandlers.ts (hunk 16 above)
Comment (santoshp210-akamai) [reply packages/manager/src/mocks/serverHandlers.ts:552]: Usually, in test files those cases are covered by sending response that might lead to errors . So we are going to test the error handling in the components

File: packages/api-v4/src/cloudpulse/services.ts (hunk 5 above)
Comment (santoshp210-akamai) [reply packages/api-v4/src/cloudpulse/services.ts]: It seems that we don't need to set additional headers unless we specifically need it for the endpoint. So I've omitted them for now, if needed as per our requirements we can add them
